File: internal/registry/lockfile.go
package registry

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

// LockfileName is namespaced to avoid colliding with the `npx skills`
// ecosystem's own lockfile semantics (spec R4.2).
const LockfileName = "skill-registry-lock.json"

// LockEntry records one installed skill. Unknown fields present on disk are
// preserved on rewrite via Extra.
type LockEntry struct {
	Slug         string   `json:"slug"`
	RegistryRef  string   `json:"registryRef,omitempty"` // slug@version
	SourceType   string   `json:"sourceType,omitempty"`
	Source       string   `json:"source,omitempty"`
	SkillPath    string   `json:"skillPath"` // repo-relative install path
	Surface      string   `json:"surface,omitempty"`
	ComputedHash string   `json:"computedHash"`
	PinnedCommit string   `json:"pinnedCommit,omitempty"`
	InstalledAt  string   `json:"installedAt,omitempty"`
	Extra        RawExtra `json:"-"`
}

// Lockfile is the on-disk document. Foreign top-level keys are preserved.
type Lockfile struct {
	Version int          `json:"version"`
	Skills  []LockEntry  `json:"skills"`
	Extra   RawExtra     `json:"-"`
	path    string       `json:"-"`
}

// RawExtra preserves unrecognized JSON object fields across a read/write cycle.
type RawExtra map[string]json.RawMessage

// LockfilePath returns the lockfile path for a repo root.
func LockfilePath(repoRoot string) string {
	return filepath.Join(repoRoot, LockfileName)
}

// LoadLockfile reads the lockfile at repoRoot, returning an empty v1 lockfile if
// none exists. Unknown fields are retained.
func LoadLockfile(repoRoot string) (*Lockfile, error) {
	path := LockfilePath(repoRoot)
	lf := &Lockfile{Version: 1, path: path, Extra: RawExtra{}}
	data, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return lf, nil
		}
		return nil, fmt.Errorf("read lockfile: %w", err)
	}
	// First pass: capture known + unknown top-level keys.
	var top map[string]json.RawMessage
	if err := json.Unmarshal(data, &top); err != nil {
		return nil, fmt.Errorf("parse lockfile %s: %w", path, err)
	}
	if raw, ok := top["version"]; ok {
		_ = json.Unmarshal(raw, &lf.Version)
		delete(top, "version")
	}
	if raw, ok := top["skills"]; ok {
		if err := lf.decodeSkills(raw); err != nil {
			return nil, err
		}
		delete(top, "skills")
	}
	lf.Extra = top
	if lf.Version == 0 {
		lf.Version = 1
	}
	return lf, nil
}

func (lf *Lockfile) decodeSkills(raw json.RawMessage) error {
	var rawEntries []map[string]json.RawMessage
	if err := json.Unmarshal(raw, &rawEntries); err != nil {
		return fmt.Errorf("parse lockfile skills: %w", err)
	}
	known := map[string]bool{
		"slug": true, "registryRef": true, "sourceType": true, "source": true,
		"skillPath": true, "surface": true, "computedHash": true,
		"pinnedCommit": true, "installedAt": true,
	}
	for _, rawEntry := range rawEntries {
		var entry LockEntry
		merged, _ := json.Marshal(rawEntry)
		if err := json.Unmarshal(merged, &entry); err != nil {
			return fmt.Errorf("parse lockfile entry: %w", err)
		}
		extra := RawExtra{}
		for k, v := range rawEntry {
			if !known[k] {
				extra[k] = v
			}
		}
		entry.Extra = extra
		lf.Skills = append(lf.Skills, entry)
	}
	return nil
}

// Upsert inserts or replaces an entry keyed by SkillPath.
func (lf *Lockfile) Upsert(entry LockEntry) {
	for i := range lf.Skills {
		if lf.Skills[i].SkillPath == entry.SkillPath {
			entry.Extra = mergeExtra(lf.Skills[i].Extra, entry.Extra)
			lf.Skills[i] = entry
			return
		}
	}
	lf.Skills = append(lf.Skills, entry)
}

// Remove deletes an entry by SkillPath, reporting whether one was removed.
func (lf *Lockfile) Remove(skillPath string) bool {
	for i := range lf.Skills {
		if lf.Skills[i].SkillPath == skillPath {
			lf.Skills = append(lf.Skills[:i], lf.Skills[i+1:]...)
			return true
		}
	}
	return false
}

// FindBySlug returns the first entry matching slug.
func (lf *Lockfile) FindBySlug(slug string) *LockEntry {
	for i := range lf.Skills {
		if strings.EqualFold(lf.Skills[i].Slug, slug) {
			return &lf.Skills[i]
		}
	}
	return nil
}

// Save writes the lockfile back, preserving foreign fields and using stable
// ordering. It only touches disk when there is content or an existing file.
func (lf *Lockfile) Save() error {
	// Sort entries by SkillPath for deterministic diffs.
	sort.SliceStable(lf.Skills, func(i, j int) bool {
		return lf.Skills[i].SkillPath < lf.Skills[j].SkillPath
	})
	out := map[string]json.RawMessage{}
	for k, v := range lf.Extra {
		out[k] = v
	}
	versionRaw, _ := json.Marshal(lf.Version)
	out["version"] = versionRaw
	skillsRaw, err := lf.encodeSkills()
	if err != nil {
		return err
	}
	out["skills"] = skillsRaw
	buf, err := json.MarshalIndent(out, "", "  ")
	if err != nil {
		return err
	}
	buf = append(buf, '\n')
	return os.WriteFile(lf.path, buf, 0o644)
}

func (lf *Lockfile) encodeSkills() (json.RawMessage, error) {
	entries := make([]map[string]json.RawMessage, 0, len(lf.Skills))
	for _, e := range lf.Skills {
		m := map[string]json.RawMessage{}
		for k, v := range e.Extra {
			m[k] = v
		}
		// Marshal known fields via a shadow struct (omitempty respected).
		knownBuf, err := json.Marshal(struct {
			Slug         string `json:"slug"`
			RegistryRef  string `json:"registryRef,omitempty"`
			SourceType   string `json:"sourceType,omitempty"`
			Source       string `json:"source,omitempty"`
			SkillPath    string `json:"skillPath"`
			Surface      string `json:"surface,omitempty"`
			ComputedHash string `json:"computedHash"`
			PinnedCommit string `json:"pinnedCommit,omitempty"`
			InstalledAt  string `json:"installedAt,omitempty"`
		}{e.Slug, e.RegistryRef, e.SourceType, e.Source, e.SkillPath, e.Surface, e.ComputedHash, e.PinnedCommit, e.InstalledAt})
		if err != nil {
			return nil, err
		}
		var knownMap map[string]json.RawMessage
		if err := json.Unmarshal(knownBuf, &knownMap); err != nil {
			return nil, err
		}
		for k, v := range knownMap {
			m[k] = v
		}
		entries = append(entries, m)
	}
	return json.Marshal(entries)
}

func mergeExtra(base, overlay RawExtra) RawExtra {
	if base == nil && overlay == nil {
		return nil
	}
	merged := RawExtra{}
	for k, v := range base {
		merged[k] = v
	}
	for k, v := range overlay {
		merged[k] = v
	}
	return merged
}

File: crates/skillrank-core/src/lockfile.rs
//! Repo-root lockfile recording installed skills. Preserves foreign fields (top
//! level and per entry) so it can be co-owned by other tooling.

use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};
use std::path::{Path, PathBuf};

/// Namespaced to avoid colliding with other ecosystems' lockfiles.
pub const LOCKFILE_NAME: &str = "skill-registry-lock.json";

/// One installed skill. Unknown fields on disk are preserved via `extra`.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct LockEntry {
    pub slug: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub registry_ref: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub source_type: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub source: String,
    pub skill_path: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub surface: String,
    pub computed_hash: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub pinned_commit: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub installed_at: String,
    #[serde(flatten)]
    pub extra: Map<String, Value>,
}

/// The on-disk lockfile document. Foreign top-level keys are preserved in `extra`.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Lockfile {
    pub version: i32,
    #[serde(default)]
    pub skills: Vec<LockEntry>,
    #[serde(flatten)]
    pub extra: Map<String, Value>,
    #[serde(skip)]
    path: PathBuf,
}

impl Lockfile {
    /// The lockfile path for a repo root.
    pub fn path_for(repo_root: &Path) -> PathBuf {
        repo_root.join(LOCKFILE_NAME)
    }

    /// Read the lockfile at repo_root, or return an empty v1 lockfile if none exists.
    pub fn load(repo_root: &Path) -> std::io::Result<Lockfile> {
        let path = Self::path_for(repo_root);
        match std::fs::read_to_string(&path) {
            Ok(data) if !data.trim().is_empty() => {
                let mut lf: Lockfile = serde_json::from_str(&data).map_err(|e| {
                    std::io::Error::new(
                        std::io::ErrorKind::InvalidData,
                        format!("parse lockfile {}: {e}", path.display()),
                    )
                })?;
                lf.path = path;
                if lf.version == 0 {
                    lf.version = 1;
                }
                Ok(lf)
            }
            _ => Ok(Lockfile {
                version: 1,
                skills: Vec::new(),
                extra: Map::new(),
                path,
            }),
        }
    }

    /// Insert or replace an entry keyed by skill_path, preserving foreign fields.
    pub fn upsert(&mut self, mut entry: LockEntry) {
        if let Some(existing) = self
            .skills
            .iter_mut()
            .find(|e| e.skill_path == entry.skill_path)
        {
            for (k, v) in existing.extra.clone() {
                entry.extra.entry(k).or_insert(v);
            }
            *existing = entry;
            return;
        }
        self.skills.push(entry);
    }

    /// Delete an entry by skill_path; reports whether one was removed.
    pub fn remove(&mut self, skill_path: &str) -> bool {
        let before = self.skills.len();
        self.skills.retain(|e| e.skill_path != skill_path);
        self.skills.len() != before
    }

    /// Return the first entry matching slug (case-insensitive).
    pub fn find_by_slug(&self, slug: &str) -> Option<&LockEntry> {
        self.skills
            .iter()
            .find(|e| e.slug.eq_ignore_ascii_case(slug))
    }

    /// Write the lockfile back with stable ordering, preserving foreign fields.
    pub fn save(&mut self) -> std::io::Result<()> {
        self.skills.sort_by(|a, b| a.skill_path.cmp(&b.skill_path));
        let json = serde_json::to_string_pretty(self)
            .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e))?;
        std::fs::write(&self.path, format!("{json}\n"))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn tmpdir(tag: &str) -> PathBuf {
        let d = std::env::temp_dir().join(format!("skillrank-{tag}-{}", std::process::id()));
        std::fs::create_dir_all(&d).unwrap();
        d
    }

    #[test]
    fn round_trip_preserves_foreign_fields() {
        let dir = tmpdir("lf");
        let seed = r#"{"version":1,"toolMeta":{"writtenBy":"other"},"skills":[{"slug":"owner/foreign","skillPath":".claude/skills/foreign/SKILL.md","computedHash":"sha256:aaa","customField":"keep"}]}"#;
        std::fs::write(dir.join(LOCKFILE_NAME), seed).unwrap();

        let mut lf = Lockfile::load(&dir).unwrap();
        lf.upsert(LockEntry {
            slug: "owner/ours".into(),
            skill_path: ".claude/skills/ours/SKILL.md".into(),
            computed_hash: "sha256:bbb".into(),
            source_type: "github".into(),
            ..Default::default()
        });
        lf.save().unwrap();

        let raw = std::fs::read_to_string(dir.join(LOCKFILE_NAME)).unwrap();
        assert!(raw.contains("toolMeta"), "foreign top-level key dropped");
        assert!(raw.contains("customField"), "foreign per-entry key dropped");

        let reloaded = Lockfile::load(&dir).unwrap();
        assert_eq!(reloaded.skills.len(), 2);
        std::fs::remove_dir_all(&dir).ok();
    }

    #[test]
    fn remove_reports_correctly() {
        let dir = tmpdir("lf-rm");
        let mut lf = Lockfile::load(&dir).unwrap();
        lf.upsert(LockEntry { slug: "a".into(), skill_path: "p/a".into(), computed_hash: "h".into(), ..Default::default() });
        assert!(lf.remove("p/a"));
        assert!(!lf.remove("p/a"));
        std::fs::remove_dir_all(&dir).ok();
    }
}
